proposals/0292-package-registry-service.md

```diff
@@ -38,77 +43,222 @@ it's not well-suited to this kind of workflow for the following reasons:
-| ------ | ---------------------------------------------------- | ------------------------------------------------ |
-| `GET`  | `/{package}`                                         | List package releases                            |
-| `GET`  | `/{package}/{version}`                               | Fetch metadata for a package release             |
-| `GET`  | `/{package}/{version}/Package.swift{?swift-version}` | Fetch manifest for a package release             |
-| `GET`  | `/{package}/{version}.zip`                           | Download source archive for a package release    |
-| `PUT`  | `/{package}/{version}{?commit,branch,tag,path,url}`  | Publish a package release or update its metadata |
+| Method | Path                                                      | Description                                     |
+| ------ | --------------------------------------------------------- | ----------------------------------------------- |
+| `GET`  | `/{scope}/{name}`                                         | List package releases                           |
+| `GET`  | `/{scope}/{name}/{version}`                               | Fetch metadata for a package release            |
+| `GET`  | `/{scope}/{name}/{version}/Package.swift{?swift-version}` | Fetch manifest for a package release            |
+| `GET`  | `/{scope}/{name}/{version}.zip`                           | Download source archive for a package release   |
+| `GET`  | `/identifiers{?url}`                                      | Lookup package identifiers registered for a URL |
 
 A formal specification for the package registry interface
 is provided alongside this proposal.
 In addition,
-an OpenAPI (v3) document and a reference implementation
+an OpenAPI (v3) document
+and a reference implementation written in Swift
 are provided for the convenience of developers interested
 in building their own package registry.
 
 ### Changes to Swift Package Manager
 
+#### Package identity
+
+Currently, the identity of a package is computed from
+the last path component of its effective URL
+(which can be changed with dependency mirroring).
+However, this approach can lead to a conflation of
+distinct packages with similar names
+and the duplication of the same package under different names.
+
+We propose using a scoped identifier
+in the form `scope.package-name`
+to identify package dependencies.
+
+A *scope* provides a namespace for related packages within a package registry.
+A package scope consists of
+alphanumeric characters and hyphens.
+Hyphens may not occur at the beginning or end,
+nor consecutively within a scope.
+The maximum length of a package name is 39 characters.
+A valid package scope matches the following regular expression pattern:
+
+```regexp
+\A[a-zA-Z\d](?:[a-zA-Z\d]|-(?=[a-zA-Z\d])){0,39}\z
+```
+
+A package's *name* is specified by the `name` provided in its manifest.
```

Comment:
@mattt a design direction we have been advancing towards is removing the name from the manifest, since it has no functional utility and is causing confusion given lack of uniqueness. does this pose a problem in this context? asked differently, if we had no name field in the manifest, would it impact the design in any way?

Reply:
Removing the `name` field in a package wouldn't substantively affect this proposal. The only necessary change would be to say remove any mention of a package's name being defined in its manifest. If the `name` field were to become deprecated or decouple from package identity, I'd want to tighten up the naming rules to adopt the same pattern as orgs and not use `XID_START`/`XID_CONTINUE`, which were chosen primarily to accommodate existing package names in the ecosystem.

Reply:
I'd push back on the claim that the `name` field has no functional utility. It's true that, with a robust identity scheme, a package's name isn't necessary for resolving a package dependency graph. However, I think the `name` field provides utility in other ways. Among other things, a name gives the UI a human-readable label and guards against intentional or unintentional duplication of packages across multiple identities.

Reply:
Another side effect is we won't be able to provide reasonable default for `archive-source` subcommand?

Reply:
@yim-lee That's a great point. Yes — that would be another side-effect of removing `name`.

Reply:
@mattt can you elaborate on " guards against intentional or unintentional duplication of packages across multiple identities."

Reply:
also we don't have to get into removing on not-removing the name attribute from the manifest, but rewording "A package's *name* is specified by the `name` provided in its manifest." may be useful, i.e. "The package name, which must be unique within the scope"